DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/differentiate_with.jl

```diff
@@ -10,31 +11,41 @@ function Mooncake.rrule!!(
 
     # output is a vector, so we need to use the vector pullback
     function pullback_array!!(dy::NoRData)
-        tx = DI.pullback(f, backend, primal_x, (fdata(y.dx),))
-        @assert only(tx) isa rdata_type(typeof(primal_x))
-        return NoRData(), only(tx)
+        tx = DI.pullback(f, backend, primal_x, (y.dx,))
+        @assert rdata(only(tx)) isa rdata_type(tangent_type(typeof(primal_x)))
+        return NoRData(), rdata(only(tx))
     end
 
     # output is a scalar, so we can use the scalar pullback
     function pullback_scalar!!(dy::Number)
         tx = DI.pullback(f, backend, primal_x, (dy,))
-        @assert only(tx) isa rdata_type(typeof(primal_x))
-        return NoRData(), only(tx)
+        @assert rdata(only(tx)) isa rdata_type(tangent_type(typeof(primal_x)))
+        return NoRData(), rdata(only(tx))
     end
 
     # output is a Tuple, NTuple
     function pullback_tuple!!(dy::Tuple)
         tx = DI.pullback(f, backend, primal_x, (dy,))
-        @assert only(tx) isa rdata_type(typeof(primal_x))
-        return NoRData(), only(tx)
+        @assert rdata(only(tx)) isa rdata_type(tangent_type(typeof(primal_x)))
+        return NoRData(), rdata(only(tx))
     end
 
-    pullback = if typeof(primal(y)) <: Number
+    # inputs are non Differentiable
+    function pullback_nodiff!!(dy::NoRData)
+        @assert tangent_type(typeof(primal(x))) <: NoTangent
+        return NoRData(), dy
+    end
+
+    pullback = if tangent_type(typeof(primal(x))) <: NoTangent
+        pullback_nodiff!!
+    elseif typeof(primal(y)) <: Number
         pullback_scalar!!
     elseif typeof(primal(y)) <: Array
         pullback_array!!
-    else
+    elseif typeof(primal(y)) <: Tuple
         pullback_tuple!!
+    else
+        error("$(typeof(primal(y))) primal type currently not supported.")
```

Comment:
Can you also print the type information of `f` and `x`, which are helpful for debugging.